resources/2023-11-application-diagnostics-prototype.md

```diff
@@ -0,0 +1,215 @@
+
+#### Configuring OpenTelemetry as part of an environment
+
+As an operator, I can configure diagnostics stores for use in environments. Then I can configure a Radius environment to use those those diagnostics stores. When I do this I am accomplishing two things: 1) making the supported diagnostics stores and their configuration explicit as part of the environment, 2) making the configuration of collectors for applications automatic based on the configuration that I provided.
+
+In our demo setup we're going to use Azure Monitor + Grafana for both metrics, and logs. We're not addressing tracing in this demo. These technologies are good choices because Grafana is self-hostable and widely used, and Azure Monitor as a backend is a good subtitute for any OTEL-compatible source. It would be easy for us to move this demo to another cloud by swapping out Azure Monitor.
+
+#### Collecting and storing diagnostics data automatically for applications
+
+As a developer, I can instrument my code using the OpenTelemetry SDK to provide metrics. I can write logs to the console using the logging packages I prefer for my application. When I deploy my application, using Radius, the configuration for OpenTelemetry will be automatically injected. I don't know have to know about the configuration of, or have access to where the diagnostics data goes.
+
+In our demo setup we're going to start by adding metrics to our tutorial/demo application that's written using Node.js/TypeScript. We'll introduce a change that causes some request to randomly fail, resulting in an HTTP 502 status code. When we place load on the application we'll see a high number of HTTP 502 failures, which will trigger an alert in the next scenario. 
+
+For the evolution of this feature it should not matter what language we use to build the application.
+
+#### Responding to an alert about application availability
+
+As an SRE, I need to support applications that I did not write, and I may not understand their architecture. I have to use the diagnostics from applications to understand each incident and recommend actions. For applications that are managed by Radius I get comprehensive metadata that helps with navigation and comprehension of the telemetry signals. I can easily identify the environment, application name, and the specific component that has issues. Then I can use the application graph to understand how the application is architected and which cloud resources are involved. This helps me identify the source of a problem as well as whom to contact.
+
+In our demo setup, we're going to create an alert for the error rate coming out of an HTTP application (error golden signal). Our SRE will use the metadata to query the application graph and understand the architecture. Then they will craft a log query that helps them identify the failure pattern. 
+
```

Comment:
Something I was thinking about here... It would be cool if we could have the recipe we're using for Redis hook up Azure Monitor as well. 

My idea goes something like this:

- SRE see the alert
- First thing they do is check the application graph to understand the architecture and see that the application is using Redis.
- They suspect that Redis is down, so they check that with an ad-hoc query. All green!
- So then they dive into logs and find the bug. 

Thoughts? /cc @AaronCrawfis @kachawla

Reply:
An additional aspect we should consider for the prototype/discovery process is the iterative nature of diagnostics. Any diagnostics solution has an inherent overhead. Typically, the diagnostics dial (for example log level in case of logging) is turned way down in production _until_ an issue surfaces. When it does, the dial gets turned higher and higher (depending on root cause). This iterative process needs to be super simple and as non intrusive as possible.

Reply:
For example, once I've identified the component that is failing via the architecture graph, even something as simple as moving a slider to up the log level would be super cool.

Reply:
This is another idea that's super cool 👍 

I think we can probably doing this in centralized way through the configuration of the collector: https://opentelemetry.io/docs/collector/configuration/#processors